#Imports

In [57]:
import pandas as pd
import missingno as msno

#Loading Data

In [58]:
df = pd.read_csv('/content/sales_predictions.csv')


In [59]:
df.info()
df.head(20)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0            FDA15        9.300          Low Fat         0.016047   
1            DRC01        5.920          Regular         0.019278   
2            FDN15       17.500          Low Fat         0.016760   
3            FDX07       19.200          Regular         0.000000   
4            NCD19        8.930          Low Fat         0.000000   
5            FDP36       10.395          Regular         0.000000   
6            FDO10       13.650          Regular         0.012741   
7            FDP10          NaN          Low Fat         0.127470   
8            FDH17       16.200          Regular         0.016687   
9            FDU28       19.200          Regular         0.094450   
10           FDY07       11.800          Low Fat         0.000000   
11           FDA03       18.500          Regular         0.045464   
12           FDX32       15.100          Regular         0.100014   
13           FDS46       17.600          Regular         0.047257   
14           FDF32       16.350          Low Fat         0.068024   
15           FDP49        9.000          Regular         0.069089   
16           NCB42       11.800          Low Fat         0.008596   
17           FDP49        9.000          Regular         0.069196   
18           DRI11          NaN          Low Fat         0.034238   
19           FDU02       13.350          Low Fat         0.102492   

                Item_Type  Item_MRP Outlet_Identifier  \
0                   Dairy  249.8092            OUT049   
1             Soft Drinks   48.2692            OUT018   
2                    Meat  141.6180            OUT049   
3   Fruits and Vegetables  182.0950            OUT010   
4               Household   53.8614            OUT013   
5            Baking Goods   51.4008            OUT018   
6             Snack Foods   57.6588            OUT013   
7             Snack Foods  107.7622            OUT027   
8            Frozen Foods   96.9726            OUT045   
9            Frozen Foods  187.8214            OUT017   
10  Fruits and Vegetables   45.5402            OUT049   
11                  Dairy  144.1102            OUT046   
12  Fruits and Vegetables  145.4786            OUT049   
13            Snack Foods  119.6782            OUT046   
14  Fruits and Vegetables  196.4426            OUT013   
15              Breakfast   56.3614            OUT046   
16     Health and Hygiene  115.3492            OUT018   
17              Breakfast   54.3614            OUT049   
18            Hard Drinks  113.2834            OUT027   
19                  Dairy  230.5352            OUT035   

    Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                        1999      Medium               Tier 1   
1                        2009      Medium               Tier 3   
2                        1999      Medium               Tier 1   
3                        1998         NaN               Tier 3   
4                        1987        High               Tier 3   
5                        2009      Medium               Tier 3   
6                        1987        High               Tier 3   
7                        1985      Medium               Tier 3   
8                        2002         NaN               Tier 2   
9                        2007         NaN               Tier 2   
10                       1999      Medium               Tier 1   
11                       1997       Small               Tier 1   
12                       1999      Medium               Tier 1   
13                       1997       Small               Tier 1   
14                       1987        High               Tier 3   
15                       1997       Small               Tier 1   
16                       2009      Medium               Tier 3   
17                       1999      Medium               Tier 1   
18                       1985      Medium               Tier 3   
19                       2004      

#Data Cleaning

#Assignment

## 1. How many rows and columns?

In [60]:
df.shape

(8523, 12)

In [61]:
print(f'There is {df.shape[0]} rows, and {df.shape[1]} coulums.')

There is 8523 rows, and 12 coulums.


## 2. What are the datatypes of each variable?

In [62]:
df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

In [63]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


## 3. Are there duplicates? If so, drop any duplicates.

In [64]:
df.duplicated().sum()

0

In [65]:
print(f'There is {df.duplicated().sum()} duplicate/s in our data.')

There is 0 duplicate/s in our data.


## 4. Identify missing values.

In [66]:
NAs = df.isna().sum()
NAs

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [67]:
df.isna().sum().sum()

3873

In [68]:
#checking percentage of NaN
NaNpercent = ((NAs / df.shape[0])*100).round(2)
NaNpercent

Item_Identifier               0.00
Item_Weight                  17.17
Item_Fat_Content              0.00
Item_Visibility               0.00
Item_Type                     0.00
Item_MRP                      0.00
Outlet_Identifier             0.00
Outlet_Establishment_Year     0.00
Outlet_Size                  28.28
Outlet_Location_Type          0.00
Outlet_Type                   0.00
Item_Outlet_Sales             0.00
dtype: float64

In [69]:
print(f'We will be dropping a total of {((df.isna().sum().sum())/len(df)*100).round(2)}% of rows if we drop all the NaNs ')

We will be dropping a total of 45.44% of rows if we drop all the NaNs 


## 5. Decide on how to address the missing values and do it! (This requires your judgement, so explain your choice).


In [70]:
#Filter Identifier

filter27 = df['Outlet_Identifier'] == 'OUT027'
filter13 = df['Outlet_Identifier'] == 'OUT013'
filter49 = df['Outlet_Identifier'] == 'OUT049'
filter46 = df['Outlet_Identifier'] == 'OUT046'
filter35 = df['Outlet_Identifier'] == 'OUT035'
filter45 = df['Outlet_Identifier'] == 'OUT045'
filter18 = df['Outlet_Identifier'] == 'OUT018'
filter17 = df['Outlet_Identifier'] == 'OUT017'
filter10 = df['Outlet_Identifier'] == 'OUT010'
filter19 = df['Outlet_Identifier'] == 'OUT019'

df[filter10].head(17)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
3             FDX07       19.200          Regular         0.000000   
28            FDE51        5.925          Regular         0.161467   
30            FDV38       19.250          Low Fat         0.170349   
45            FDM39        6.420               LF         0.089499   
65            FDC46       17.700               LF         0.195068   
90            FDW20       20.750          Low Fat         0.040421   
122           FDB14       20.250          Regular         0.171939   
133           FDS52        8.890          Low Fat         0.009163   
139           NCN07       18.500          Low Fat         0.056816   
174           FDI32       17.700          Low Fat         0.291865   
175           FDR47       17.850               LF         0.000000   
177           NCU05       11.800          Low Fat         0.098312   
205           FDF24       15.500          Regular         0.042465   
209           FDX40       12.850          Low Fat         0.165694   
214           NCZ54       14.650          Low Fat         0.000000   
215           FDE10        6.670          Regular         0.150555   
216           FDR44        6.110          Regular         0.000000   

                 Item_Type  Item_MRP Outlet_Identifier  \
3    Fruits and Vegetables  182.0950            OUT010   
28                   Dairy   45.5086            OUT010   
30                   Dairy   55.7956            OUT010   
45                   Dairy  178.1002            OUT010   
65             Snack Foods  185.4266            OUT010   
90   Fruits and Vegetables  122.1730            OUT010   
122                 Canned   92.5120            OUT010   
133           Frozen Foods  101.7016            OUT010   
139                 Others  132.1284            OUT010   
174  Fruits and Vegetables  115.1834            OUT010   
175                 Breads  196.5794            OUT010   
177     Health and Hygiene   81.4618            OUT010   
205           Baking Goods   81.5934            OUT010   
209           Frozen Foods   39.7164            OUT010   
214              Household  161.5552            OUT010   
215            Snack Foods  130.0626            OUT010   
216  Fruits and Vegetables  131.2968            OUT010   

     Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
3                         1998         NaN               Tier 3   
28                        1998         NaN               Tier 3   
30                        1998         NaN               Tier 3   
45                        1998         NaN               Tier 3   
65                        1998         NaN               Tier 3   
90                        1998         NaN               Tier 3   
122                       1998         NaN               Tier 3   
133                       1998         NaN               Tier 3   
139                       1998         NaN               Tier 3   
174                       1998         NaN               Tier 3   
175                       1998         NaN               Tier 3   
177                       1998         NaN               Tier 3   
205                       1998         NaN               Tier 3   
209                       1998         NaN               Tier 3   
214                       1998         NaN               Tier 3   
215                       1998         NaN               Tier 3   
216                       1998         NaN               Tier 3   

       Outlet_Type  Item_Outlet_Sales  
3    Grocery Store           732.3800  
28   Grocery Store           178.4344  
30   Grocery Store           163.7868  
45   Grocery Store           358.2004  
65   Grocery Store           184.4266  
90   Grocery Store           369.5190  
122  Grocery Store           186.4240  
133  Grocery Store           101.2016  
139  Grocery Store           263.6568  
174  Grocery Store           345.5502  
175  Grocery Store           585.2382  
177  Grocer

In [71]:
df.describe(include="number")

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  7060.000000      8523.000000  8523.000000                8523.000000   
mean     12.857645         0.066132   140.992782                1997.831867   
std       4.643456         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       8.773750         0.026989    93.826500                1987.000000   
50%      12.600000         0.053931   143.012800                1999.000000   
75%      16.850000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

In [72]:
df['Item_Weight'].fillna(df['Item_Weight'].mean(), inplace = True)
df.isna().sum()

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

###  Missing Values on Item_Weight:
  - Dropping Missing values on Item_Weight would mean losing 17.17% of my rows. 
  - Only "OUT019" and "OUT027" on Outlet_Identifier doesn't have weight on them

I decided to put the mean value of those Null rows in the Item_Weight because 50% of the data is actually in the range of the mean value. Dropping 17.17% rows is a great loss on the data, as well as I cannot afford to lose 2 types of Outlet_Identifier. 

In [73]:
df['Outlet_Size'].fillna('Unknown',inplace = True)
df.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

### - Missing Values on Outlet_Size
  - 28.28% of rows are Missing.
  - "OUT010", "OUT045" and "OUT017" on Outlet_Identifier doesn't have values on Outlet_Size.

we cannot afford to lose 28.28% of our data, as well as, 3 types of out Outlet_Identifier. So to save the data, I decided to put 'Unknown' on the Null values under Outlet_Size as there is no way for us to determine the size of the outlet with the current data that we have. 

## 6. Confirm that there are no missing values after addressing them.

In [74]:
print(f'The total number of missing values in our data is {df.isna().sum().sum()}.')

The total number of missing values in our data is 0.


## 7. Find and fix any inconsistent categories of data (example: fix cat, Cat, and cats so that they are consistent).

###Columns

In [75]:
df.columns

Index(['Item_Identifier', 'Item_Weight', 'Item_Fat_Content', 'Item_Visibility',
       'Item_Type', 'Item_MRP', 'Outlet_Identifier',
       'Outlet_Establishment_Year', 'Outlet_Size', 'Outlet_Location_Type',
       'Outlet_Type', 'Item_Outlet_Sales'],
      dtype='object')

In [76]:
for column in object_data_types.index:
    print(column)
    print(df[column].unique())
    print('\n')

Item_Identifier
['FDA15' 'DRC01' 'FDN15' ... 'NCF55' 'NCW30' 'NCW05']


Item_Fat_Content
['Low Fat' 'Regular' 'low fat' 'LF' 'reg']


Item_Type
['Dairy' 'Soft Drinks' 'Meat' 'Fruits and Vegetables' 'Household'
 'Baking Goods' 'Snack Foods' 'Frozen Foods' 'Breakfast'
 'Health and Hygiene' 'Hard Drinks' 'Canned' 'Breads' 'Starchy Foods'
 'Others' 'Seafood']


Outlet_Identifier
['OUT049' 'OUT018' 'OUT010' 'OUT013' 'OUT027' 'OUT045' 'OUT017' 'OUT046'
 'OUT035' 'OUT019']


Outlet_Size
['Medium' 'Unknown' 'High' 'Small']


Outlet_Location_Type
['Tier 1' 'Tier 3' 'Tier 2']


Outlet_Type
['Supermarket Type1' 'Supermarket Type2' 'Grocery Store'
 'Supermarket Type3']




###Object Columns

In [77]:
data_types = df.dtypes
object_data_types = data_types[(data_types == "object")]
object_data_types

Item_Identifier         object
Item_Fat_Content        object
Item_Type               object
Outlet_Identifier       object
Outlet_Size             object
Outlet_Location_Type    object
Outlet_Type             object
dtype: object

In [78]:
object_data_types.index

Index(['Item_Identifier', 'Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

####Item_Identifier Column

In [79]:
df['Item_Identifier'].unique()

array(['FDA15', 'DRC01', 'FDN15', ..., 'NCF55', 'NCW30', 'NCW05'],
      dtype=object)

In [80]:
df['Item_Identifier'].value_counts()

FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64

####Item_Type Column

In [81]:
df['Item_Type'].unique()

array(['Dairy', 'Soft Drinks', 'Meat', 'Fruits and Vegetables',
       'Household', 'Baking Goods', 'Snack Foods', 'Frozen Foods',
       'Breakfast', 'Health and Hygiene', 'Hard Drinks', 'Canned',
       'Breads', 'Starchy Foods', 'Others', 'Seafood'], dtype=object)

In [82]:
df['Item_Type'].value_counts()

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

####Item_Fat_Content Column

In [83]:
df['Item_Fat_Content'].unique()

array(['Low Fat', 'Regular', 'low fat', 'LF', 'reg'], dtype=object)

In [84]:
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [85]:
df['Item_Fat_Content'].replace({'LF':'Low Fat', 'low fat': 'Low Fat', 'reg':'Regular' }, inplace=True)
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

#### Outlet_Identifier Column

In [86]:
df['Outlet_Identifier'].unique()

array(['OUT049', 'OUT018', 'OUT010', 'OUT013', 'OUT027', 'OUT045',
       'OUT017', 'OUT046', 'OUT035', 'OUT019'], dtype=object)

In [87]:
df['Outlet_Identifier'].value_counts()

OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64

####Outlet_Size Column

In [88]:
df['Outlet_Size'].unique()

array(['Medium', 'Unknown', 'High', 'Small'], dtype=object)

In [89]:
df['Outlet_Size'].value_counts()

Medium     2793
Unknown    2410
Small      2388
High        932
Name: Outlet_Size, dtype: int64

####Outlet_Location_Type Column

In [90]:
df['Outlet_Location_Type'].unique()

array(['Tier 1', 'Tier 3', 'Tier 2'], dtype=object)

In [91]:
df['Outlet_Location_Type'].value_counts()

Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: Outlet_Location_Type, dtype: int64

####Outlet_Type Column

In [92]:
df['Outlet_Type'].unique()

array(['Supermarket Type1', 'Supermarket Type2', 'Grocery Store',
       'Supermarket Type3'], dtype=object)

In [93]:
df['Outlet_Type'].value_counts()

Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: Outlet_Type, dtype: int64

###8. For any numerical columns, obtain the summary statistics of each (min, max, mean).


In [94]:
df.describe(include="number")

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  8523.000000      8523.000000  8523.000000                8523.000000   
mean     12.857645         0.066132   140.992782                1997.831867   
std       4.226124         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       9.310000         0.026989    93.826500                1987.000000   
50%      12.857645         0.053931   143.012800                1999.000000   
75%      16.000000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

* Here are a few resources to help you consider more advanced ways to deal with missing values in your dataset (check out the optional additional Pandas resources at the end of this Chapter for more resources on this):
- Missing Data, its types, and statistical methods to deal with it.
- 6 Different Ways to Compensate for Missing Values In a Dataset.
- Dealing with missing data: Key assumptions and methods for applied analysis.
- Feel free to use any other resources or techniques you find online that you think are appropriate/useful.